### ISSUE TITLE
Dimensions of Embedding Layer

### ISSUE THREAD

**Issue by Tuyki** [fchollet/keras#631]
Greetings, 

I am now learning Keras and find it an excellent framework! 
I also have a question regarding the Embedding layer. Although I am aware of the way an embedding layer works, I do not yet quite understand the input/output dimensions of it in Keras. 
Here's a toy example of mine: Say, I have 10 entities, for each of which I need a 5D dense representation. And out of the representations I want to calculate an output. (which doesn't make any sense I just want to understand the way how it works)
So the codes are: 

model = Sequential()
model.add(Embedding(10, 5))
model.add(Dense(5, 1, init='uniform', activation = 'linear', W_regularizer = l2(0.01)))
sgd = SGD(lr = 0.1, decay = 0.01, momentum = 0.5, nesterov = False)
model.compile(loss='mean_squared_error', optimizer=sgd)

x = np.random.choice(np.arange(10), 10, replace = False).reshape(10, 1)
y = np.random.randn(10).reshape(10, 1)

model.fit(x, y, nb_epoch=100, batch_size=3)

so the error is: 
"TypeError: ('Bad input argument to theano function with name "build/bdist.linux-x86_64/egg/keras/models.py:400"  at index 1(0-based)', 'Wrong number of dimensions: expected 3, got 2 with shape (3, 1).')"

As stated in the documentation, the output of embedding layer is a 3D tensor of shape (nb_samples, maxlen, output_dim). How should I reshape (.add(Reshape(10, 5)) doesn't work for me) the output of the embedding layer? And what does this "maxlen" stand for? 

Many thanks
Kind regards, 

T

**Comment by fchollet** [comment]
`maxlen` is the maximum length of the sequences you are embedding. So it's
the 2nd dimension (time dimension) in the output tensor.

You cannot use a `Dense` layer after an `Embedding` layers, because the
Embedding layer is meant to embed sequences (represented as 3D tensors).
Dense works on 2D tensors (samples, features).

> Say, I have 10 entities, for each of which I need a 5D dense
representation.

It's not entirely clear what you mean. But taking it literally, you could
simply use as your initial layer `Dense(10, 5)`. The input vectors would be
all-zero except for a 1 at the location of the entity index (eg. [0, 0, 1,
0, 0, 0, 0, 0, 0, 0] for entity 3).

If you don't have sequences, don't use Embeddings.

**Comment by Tuyki** [comment]
Hello @fchollet , 

thank you very much for your answers! 
Originally, I did implement the embedding layer using the binary vector you suggested. But then I suddenly realized that there are two drawbacks of doing so: 
- First, I cannot deactivate the bias, which doesn't seem to fit into the embedding concepts, unless I set a very large b_regularizer; 
- Second, since keras doesn't yet support sparse input (correct me if I'm wrong :P ), I assume it wouldn't be so efficient to have those binary vectors; 
So that's why I became interested in the Embedding layer. 

Interestingly, I managed to build a Dense layer with input from a Embedding layer by inserting a Flatten:

model = Sequential()
model.add(Embedding(10, 5))
model.add(Flatten())
model.add(Dense(5, 1, init='uniform', activation = 'linear', W_regularizer = l2(0.01)))
sgd = SGD(lr = 0.1, decay = 0.01, momentum = 0.5, nesterov = False)
model.compile(loss='mean_squared_error', optimizer=sgd)

x = np.random.choice(np.arange(10), 10, replace = False).reshape(10, 1)
y = np.random.randn(10).reshape(10, 1)

model.fit(x, y, nb_epoch=100, batch_size=3)

But I'm not sure if that is correct, although according to Documentation it seems to be... 

To summarize: 
1) Under certain circumstances Embedding layers might also be helpful for non-sequence data (e.g. I'm working on knowledge graphs) and it'd be hence great if there be some direct interface between it and Dense. 
2) Would it be possible to add the functionality of deactivating bias in each layer?
3) Would it also be possible to support sparse input of a Keras model? 

Thank you very much for your attention and answers

Best regards, 

T

**Comment by ndronen** [comment]
I agree, @Tuyki: embeddings do not imply order.  It should be possible to use them with unordered inputs (e.g. bag-of-words models in NLP, knowledge graphs).  I believe both chainer and torch allow this.

See PRs #653 and #684 for functionality that will allow you to use embeddings in that way.  If you need the functionality right away, just clone and install jmhessel's keras repository.

**Comment by Sandy4321** [comment]
new drop out for embeddings layer : conventional drop outs leads to overfititng
http://arxiv.org/abs/1512.05287
A Theoretically Grounded Application of Dropout in Recurrent Neural Networks

Yarin Gal

**Comment by volkanagun** [comment]
I have problem with the shape of embedding layer.
The error is expected embedding_1 to have 3 dimensions, but got array with shape (40, 2)
The code is below...



```
layer = Embedding(maxFeatures, output_dim=dimSize, input_length=2, init="uniform")
model = Sequential()
model.add(layer)
model.compile(loss='mse', optimizer='rmsprop')
sampling_table = sequence.make_sampling_table(maxFeatures)

print('Tokenize...')
print("Fit tokenizer...")
tokenizer = text.Tokenizer(nb_words=maxFeatures)
tokenizer.fit_on_texts(LineIterator(name))
for i, seq in enumerate(tokenizer.texts_to_sequences_generator(LineIterator(name))):
     couples, labels = sequence.skipgrams(seq, maxFeatures, window_size=4, negative_samples=1.,        sampling_table=sampling_table)
        instances = np.array(couples, dtype="int32")
        cats = np_utils.to_categorical(labels,2) 
        model.fit(instances,cats)
        samplesSeen += len(labels)
```

**Comment by Xnsam** [comment]
@volkanagun Hi, did you figure out your problem's solution?...
I have the same issue...

**Comment by Pratham9811** [comment]
@Xnsam did you find solution? I'm facing same issue
